doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,157 @@
+
+### Observed problems with changing key names
+
+Suppose there is a plugin that changes key names.
+It converts to lowercase for the runtime representation, and to uppercase for the stored representation.
+The plugin is executed in the post-storage phase for the `get` operation and the pre-storage phase for the `set` operation.
+This results in the keys being in UPPERCASE in the configuration files, but they are presented in lowercase to other plugins and applications using the Elektra API.
+
+For example, here is a configuration file with a hypothetical format:
+
+```
+/DISPLAY/BRIGHTNESS = 100
+```
+
+As can be seen, the keys are in UPPERCASE within the configuration file.
+In Elektra keys are case-sensitive.
+As operations on keysets such as `ksLookup` operate with _runtime_ data after the post-storage phase, `kdb get /DISPLAY/BRIGHTNESS` will fail.
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
+In this example, the notification plugins will receive all-UPPERCASE keys, and send out notifications with those all-UPPERCASE keys.
+An application listening to those notifications will not be able to query Elektra for those keys, as for Elektra those UPPERCASE keys do not exist.
+
+Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
+This is because key names are read-only when they are contained in a `KeySet`.
+In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
+The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
+As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.
+
+### Observed problems with changing values (normalization)
+
+Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
+If a plugin does change tracking, this will lead to false positives.
+
+```
+user:/limits/openfiles = 1
+```
+
+If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
+The value-changing plugin, however, will reset that value back to `1`.
+So in practice, the configuration has not been changed.
+Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.
+
+## Constraints
+
+1. We want to enable some kind of transformations
+2. Renaming keys should be possible, at least in storage plugins
```

Comment:
Adding "at least in storage plugins" makes this constraint meaningless. Before the end of the `get/storage` phase, `libelektra-kdb` doesn't know about any keys. So storage plugins can do whatever transformations they want to. Any such transformation would be considered part of the semantic translation between the storage format and Elektra.


I guess what we want is "user-defined renaming" without the need to build the whole functionality into every storage plugin separately. I.e. either have separate renaming plugins, or have a generic way through which _all_ storage plugins can support renaming.